src/backend/mpi/swdsm.cpp

```diff
@@ -436,315 +433,264 @@ void handler(int sig, siginfo_t *si, void *unused){
-		MPI_Get_accumulate(&id, 1, MPI_LONG, &tempsharer, 1, MPI_LONG,
-			homenode, classidx, 1, MPI_LONG, MPI_BOR, sharerWindow);
-		MPI_Get(&tempwriter, 1,MPI_LONG,homenode,classidx+1,1,MPI_LONG,sharerWindow);
-		MPI_Win_unlock(homenode, sharerWindow);
-	}
-
-	MPI_Win_lock(MPI_LOCK_EXCLUSIVE, workrank, 0, sharerWindow);
-	globalSharers[classidx] |= tempsharer;
-	globalSharers[classidx+1] |= tempwriter;
-	MPI_Win_unlock(workrank, sharerWindow);
-
-	unsigned long offset = getOffset(lineAddr);
-	if(isPowerOf2((tempsharer)&invid) && prevsharer == 0){ //Other private. but may not have loaded page yet.
-		unsigned long ownid = tempsharer&invid; // remove own bit
-		unsigned long owner = invalid_node; // initialize to failsafe value
-		for(n=0; n<numtasks; n++) {
-			if(1ul<<n == ownid) {
-				owner = n; //just get rank...
-				break;
+	/* Finally, get the cache data and store it temporarily */
+	MPI_Win_lock(MPI_LOCK_SHARED, load_node , 0, globalDataWindow[load_node]);
+	MPI_Get(temp_data.data(), fetch_size, cacheblock,
+					load_node, load_offset, fetch_size, cacheblock, globalDataWindow[load_node]);
+	MPI_Win_unlock(load_node, globalDataWindow[load_node]);
+
+	/* Update the cache */
+	for(std::size_t idx = start_index, p = 0; idx < end_index; idx+=CACHELINE, p+=CACHELINE){
+		/* Update only the pages necessary */
+		if(pages_to_load[p]){
+			/* Insert the data in the node cache */
+			memcpy(&cacheData[idx*block_size], &temp_data[p*block_size], block_size);
+
+			const std::size_t temp_addr = aligned_access_offset + p*block_size;
+			void* temp_ptr = static_cast<char*>(startAddr) + temp_addr;
+
+			/* If this is the first time inserting in to this index, perform vm map */
+			if(cacheControl[idx].tag == GLOBAL_NULL){
+				vm::map_memory(temp_ptr, block_size, pagesize*idx, PROT_READ);
+				cacheControl[idx].tag = temp_addr;
+			}else{
+				/* Else, just mprotect the region */
+				mprotect(temp_ptr, block_size, PROT_READ);
 			}
+			touchedcache[idx] = 1;
+			cacheControl[idx].state = VALID;
+			cacheControl[idx].dirty=CLEAN;
 		}
-		if(owner != invalid_node) {
-			MPI_Win_lock(MPI_LOCK_EXCLUSIVE, owner, 0, sharerWindow);
-			MPI_Accumulate(&id, 1, MPI_LONG, owner, classidx, 1, MPI_LONG, MPI_BOR, sharerWindow);
-			MPI_Win_unlock(owner, sharerWindow);
-		}
-
-	}
-
-	MPI_Win_lock(MPI_LOCK_SHARED, homenode , 0, globalDataWindow[homenode]);
-	MPI_Get(&cacheData[startidx*pagesize], 1, cacheblock, homenode,
-		offset, 1, cacheblock, globalDataWindow[homenode]);
-	MPI_Win_unlock(homenode, globalDataWindow[homenode]);
-
-
-	if(cacheControl[startidx].tag == GLOBAL_NULL){
-		vm::map_memory(lineptr, blocksize, pagesize*startidx, PROT_READ);
-		cacheControl[startidx].tag = lineAddr;
 	}
-	else{
-		mprotect(lineptr,pagesize*CACHELINE,PROT_READ);
-	}
-
-	touchedcache[startidx] = 1;
-	cacheControl[startidx].state = VALID;
-	cacheControl[startidx].dirty=CLEAN;
+	/* Ensure to free space occupied by tempData */
```

Comment:
`tempData` is not used anywhere else. What is this comment about?

Reply:
It was not supposed to be there at all, and no longer is!